langlib/langlib-test/src/test/resources/test-src/valuelib_test.bal

```diff
@@ -2305,6 +2305,47 @@ function testCloneWithTypeWithTuples() returns error? {
     assertFalse(x is Array);
 }
 
+type Assertion [string, string];
+
+type BoundAssertion ["let", int];
+
+type UnionTuple Assertion|BoundAssertion;
+
+type Table1 table<map<int>>;
+
+type Table2 table<map<string>>;
+
+type UnionTable Table1|Table2;
+
+function testCloneWithTypeToUnionOfTypeReference() {
+    anydata[] arrValue = ["let", 3];
+
+    BoundAssertion|error t1 = arrValue.cloneWithType(); 
+    assertFalse(t1 is error);
+    assertTrue(t1 == ["let", 3]);
+
+    Assertion|BoundAssertion|error t2 = arrValue.cloneWithType(); 
+    assertFalse(t2 is error);
+    assertTrue(t2 is BoundAssertion);
+    assertTrue(t2 == ["let", 3]);
+
+    UnionTuple|error t3 = arrValue.cloneWithType(); 
+    assertFalse(t3 is error);
+    assertTrue(t3 is BoundAssertion);
+    assertTrue(t3 == ["let", 3]);
+
+    table<map<anydata>> tab = table [{a: "a", b: "b"}];
+    Table1|Table2|error t4 = tab.cloneWithType(); 
+    assertFalse(t4 is error);
+    assertTrue(t4 is Table2);
+    assertTrue(t4 == table [{a: "a", b: "b"}]);
+
+    UnionTable|error t5 = tab.cloneWithType(); 
+    assertFalse(t5 is error);
+    assertTrue(t5 is Table2);
+    assertTrue(t5 == table [{a: "a", b: "b"}]);
+}
+
```

Comment:
```bal
function testFromJsonWithTypeToUnionOfTypeReference() {
    json[] arrValue = ["let", 3];

    BoundAssertion|error t1 = arrValue.fromJsonWithType(); 
    io:println(t1 is error); // false

    Assertion|BoundAssertion|error t2 = arrValue.fromJsonWithType(); 
    io:println(t2 is error); //  false

    UnionTuple|error t3 = arrValue.fromJsonWithType(); 
    io:println(t3 is error); //  false

    map<json>[] tab = [{a: "a", b: "b"}];
    Table1|Table2|error t4 = tab.fromJsonWithType(); 
    io:println(t4 is error);  // true - should be false

    UnionTable|error t5 = tab.fromJsonWithType(); 
    io:println(t5 is error); // true - should be false
}
```
The later 2 are still incorrect. Shall we also fix that?

Reply:
This is due to the missing validation for table value conversion. 
This is failing in the `2201.3.x` versions as well.
I'll address this in a separate issue https://github.com/ballerina-platform/ballerina-lang/issues/39385

Reply:
Added the passing cases in tests